File: solutions/46/sol.ts
const isComposite: (n: number) => boolean = (n) => {
  const squareRoot = Math.floor(Math.sqrt(n));
  for (let num = 2; num <= squareRoot; num++) {
    if (n % num === 0)
      return true;
  }
  return false;
}

const isPerfectSquare: (n: number) => boolean = (n) => {
  const squareRoot = Math.floor(Math.sqrt(n));
  return squareRoot * squareRoot === n;
}

const getEquation: (n: number, primeNumbers: number[]) => string = (n, primeNumbers) => {
  for (const primeNumber of primeNumbers) {
    let result = n - primeNumber;
    if (result % 2 === 0) {
      result = Math.floor(result / 2);
      if (isPerfectSquare(result)) {
        return `${n} = ${primeNumber} + 2 x ${result} x ${result}`
      }
    }
  }
  return ''
}

const getAnswer: () => number = () => {
  let num = 5;
  const primeNumbers = [2, 3]
  while (true) {
    if (isComposite(num)) {
      const equation = getEquation(num, primeNumbers);
      if (equation) {
        console.log(equation)
      } else {
        return num;
      }
    } else {
      primeNumbers.push(num)
    }
    num += 2;
  }
}

const answer = getAnswer();
console.log(answer);

File: solutions/100/sol.py
import math

# find blue with the equation (blue/total)*((blue-1)/(total-1)) = 1/2
def getBlue(total):
  a = 2
  b = -2
  c = -total*(total-1)

  discriminant = b**2 - 4*a*c
  delta = math.floor(math.sqrt(discriminant))

  if(discriminant != delta * delta):
    return -1

  blue = (-b + delta) / (2 * a)
  return math.floor(blue)

def getAnswer(total):
  while True:
    blue = getBlue(total)
    if blue == -1:
      total = total + 1
    else:
      return blue

print(getAnswer(21))
print(getAnswer(85+35))
print(getAnswer(10**12+1)) # impossible